src/circuits/gates.jl

```diff
@@ -455,3 +461,80 @@ gate(gn::String, s::Vector{<: Index}, ns::Int...; kwargs...) =
 ITensors.op(gn::GateName, ::SiteType"Qubit", s::Index...; kwargs...) =
   gate(gn, s...; kwargs...)
 
+
+
+"""
+    gate(M::Union{MPS,MPO}, gatename::String, site::Int; kwargs...)
+
+Generate a gate tensor for a single-qubit gate identified by `gatename`
+acting on site `site`, with indices identical to a reference state `M`.
+"""
+function gate(M::Union{MPS,MPO},
+              gatename::String,
+              site::Int; kwargs...)
+  site_ind = (typeof(M)==MPS ? siteind(M,site) :
+                               firstind(M[site], tags="Site", plev = 0))
+  return gate(gatename, site_ind; kwargs...)
+end
+
+"""
+    gate(M::Union{MPS,MPO},gatename::String, site::Tuple; kwargs...)
+
+Generate a gate tensor for a two-qubit gate identified by `gatename`
+acting on sites `(site[1],site[2])`, with indices identical to a 
+reference state `M` (`MPS` or `MPO`).
+"""
+function gate(M::Union{MPS,MPO},
+              gatename::String,
+              site::Tuple; kwargs...)
+  site_ind1 = (typeof(M)==MPS ? siteind(M,site[1]) :
+                                firstind(M[site[1]], tags="Site", plev = 0))
+  site_ind2 = (typeof(M)==MPS ? siteind(M,site[2]) :
+                                firstind(M[site[2]], tags="Site", plev = 0))
+
+  return gate(gatename,site_ind1,site_ind2; kwargs...)
+end
+
+
+gate(M::Union{MPS,MPO}, gatedata::Tuple) =
+  gate(M,gatedata...)
+
+gate(M::Union{MPS,MPO},
+     gatename::String,
+     sites::Union{Int, Tuple},
+     params::NamedTuple) =
+  gate(M, gatename, sites; params...)
+
+
+"""
+RANDOM GATE PARAMETERS
+"""
+
+randomparams(::GateName"Rx",args...; rng = Random.GLOBAL_RNG) = (θ = π*rand(rng),)
+randomparams(::GateName"Ry",args...; rng = Random.GLOBAL_RNG) = (θ = π*rand(rng),) 
+randomparams(::GateName"Rz",args...; rng = Random.GLOBAL_RNG) = (ϕ = 2*π*rand(rng),)
+randomparams(::GateName"CRz",args...;rng = Random.GLOBAL_RNG) = (ϕ = 2*π*rand(rng),)
+randomparams(::GateName"Rn",args...; rng = Random.GLOBAL_RNG) = 
+  ( θ = π * rand(rng), ϕ = 2 * π * rand(rng), λ = π * rand(rng))
+randomparams(::GateName"CRn", args...;rng = Random.GLOBAL_RNG) = 
+  ( θ = π * rand(rng), ϕ = 2 * π * rand(rng), λ = π * rand(rng))
+
+randomparams(::GateName"X",args...; kwargs...) = nothing
```

Comment:
We could just have a generic fallback function:
```julia
randomparams(::GateName, args...; kwargs...) = nothing
```
Better yet would be to have it return an empty `NamedTuple`, i.e.:
```julia
randomparams(::GateName, args...; kwargs...) = NamedTuple()
```

Reply:
Ok I did not know this was an option. I will make this change

Reply:
I think the `nothing` would be better though, since otherwise you would have all empty namedtuples showing up all around the circuit data structure:

```julia
julia> randomlayer("X",10)
10-element Array{Tuple,1}:
 ("X", 1, NamedTuple())
 ("X", 2, NamedTuple())
 ("X", 3, NamedTuple())
 ("X", 4, NamedTuple())
 ("X", 5, NamedTuple())
 ("X", 6, NamedTuple())
 ("X", 7, NamedTuple())
 ("X", 8, NamedTuple())
 ("X", 9, NamedTuple())
 ("X", 10, NamedTuple())
```

Reply:
Well, you can still check if the NameTuple() `isempty()` and not add it to the gate, just like you would have to check if it was `isnothing` if we choose `nothing` as the convention.